File: client/src/services/companyService.js
import axios from 'axios';
import { handleApiError } from '../utils/errorHandler';

const API_URL = import.meta.env.VITE_API_URL;

const companyService = {
    getCompanies: async (orderBy = 'name', orderType = 'asc', page = 1, perPage = 50) => {
        try {
            const response = await axios.get(`${API_URL}/api/companies`, {
                params: {
                    order_by: orderBy,
                    order_type: orderType,
                    page: page,
                    per_page: perPage
                },
                withCredentials: true
            });
            return response.data;
        } catch (error) {
            console.error("Error fetching companies:", error);
            throw handleApiError(error);
        }
    },

    getCompany: async (id) => {
        try {
            const response = await axios.get(`${API_URL}/api/companies/${id}`, { withCredentials: true });
            return response.data;
        } catch (error) {
            console.error("Error fetching company:", error);
            throw handleApiError(error);
        }
    },

    createCompany: async (companyData) => {
        try {
            const response = await axios.post(`${API_URL}/api/companies`, { company: companyData }, {
                withCredentials: true
            });
            return response.data;
        } catch (error) {
            if (error.response && error.response.status === 422) {
                console.error("Validation error:", error.response.data);
                throw new Error(JSON.stringify(error.response.data));
            }
            console.error("Error creating company:", error);
            throw handleApiError(error);
        }
    },

    updateCompany: async (id, companyData) => {
        try {
            const response = await axios.put(`${API_URL}/api/companies/${id}`, { company: companyData }, {
                withCredentials: true
            });
            return response.data;
        } catch (error) {
            if (error.response && error.response.status === 422) {
                console.error("Validation error:", error.response.data);
                throw new Error(JSON.stringify(error.response.data));
            }
            console.error("Error updating company:", error);
            throw handleApiError(error);
        }
    },

    deleteCompany: async (id) => {
        try {
            await axios.delete(`${API_URL}/api/companies/${id}`, { withCredentials: true });
            return { success: true };
        } catch (error) {
            console.error("Error deleting company:", error);
            throw handleApiError(error);
        }
    },

    getCompanyFields: async () => {
        try {
            const response = await axios.get(`${API_URL}/api/companies/fields`, { withCredentials: true });
            return response.data.fields;
        } catch (error) {
            console.error("Error fetching company fields:", error);
            throw handleApiError(error);
        }
    }
};

export default companyService;

File: server/app/services/freshdesk/company_service.rb
module Freshdesk
  class CompanyService
    def initialize
      @client = ApiClient.new
    end

    def get_company(id)
      begin
        @client.request(:get, "/companies/#{id}")
      rescue Freshdesk::ResourceNotFoundError
        raise Freshdesk::ResourceNotFoundError.new("Company ##{id} not found")
      rescue Freshdesk::Error => e
        raise e
      rescue => e
        raise Freshdesk::RequestError.new("Failed to get company ##{id}: #{e.message}")
      end
    end

    def create_company(company_params)
      begin
        @client.request(:post, '/companies', company_params)
      rescue Freshdesk::ValidationError => e
        raise e
      rescue Freshdesk::Error => e
        raise e
      rescue => e
        raise Freshdesk::RequestError.new("Failed to create company: #{e.message}")
      end
    end

    def update_company(id, company_params)
      begin
        @client.request(:put, "/companies/#{id}", company_params)
      rescue Freshdesk::ResourceNotFoundError
        raise Freshdesk::ResourceNotFoundError.new("Company ##{id} not found")
      rescue Freshdesk::ValidationError => e
        raise e
      rescue Freshdesk::Error => e
        raise e
      rescue => e
        raise Freshdesk::RequestError.new("Failed to update company ##{id}: #{e.message}")
      end
    end

    def delete_company(id)
      begin
        @client.request(:delete, "/companies/#{id}")
      rescue Freshdesk::ResourceNotFoundError
        raise Freshdesk::ResourceNotFoundError.new("Company ##{id} not found")
      rescue Freshdesk::Error => e
        raise e
      rescue => e
        raise Freshdesk::RequestError.new("Failed to delete company ##{id}: #{e.message}")
      end
    end

    def get_company_fields
      begin
        @client.request(:get, '/company_fields')
      rescue Freshdesk::Error => e
        raise e
      rescue => e
        raise Freshdesk::RequestError.new("Failed to get company fields: #{e.message}")
      end
    end
  end
end
